File: Python Projects/Basic Projects/Sub_Strings.py
"""This program will count how many times a set of characters appear in a string"""

class SubString:
  def __init__(self, list_words):
    """Initialize the phrase to apply the substrings method"""
    #self.phrase = phrase
    self.list_words = list_words # this one is a list 
    self.dictionary_words = dict() # this is the dictionary with count for every word in the list_words
    self.test_list = [] # this one stores the phrase truned into a list
    self.sorted_dictionary = dict() # stores the dictionary sorted alphabetically
    

  def substrings(self):
    """Search for how many times a word appears in the string"""
    self.phrase = input("Enter you phrase:").lower()

    # Creates a list of every word in the phrase
    self.test_list = self.phrase.split()
    print(self.test_list)
        
    # loop the found matches and create a dictionary counting them
    for item in self.test_list:
      for item2 in self.list_words:
        try:
          if item2 in item:
            self.dictionary_words[item2] = self.dictionary_words[item2] + 1
        except KeyError:
          self.dictionary_words[item2] = 1

    # Sorted dictionary
    
    for key, values in sorted(self.dictionary_words.items()):
      self.sorted_dictionary[key] = values

    return self.sorted_dictionary

# create a list of words to compare
dictionary = ["below","down","go","going","horn","how","howdy","it","i","low","own","part","partner","sit"]

# Instance of Substring
test_substring = SubString(dictionary)

print(test_substring.substrings())
      

File: Ruby-Projects/Basic-Projects/Sub-Strings.rb
dictionary = ["below","down","go","going","horn","how","howdy","it","i","low","own","part","partner","sit"]


def substrings(string, dictionary)

  # creates an array fro the string
  string_subarray = string.split()

  # creates a new hash
  words_hash = Hash.new()

  # loop over wach word in the string and compare it with each word in the dictionary array. Each word is converted to lowercase for the comparison
  string_subarray.each do |word|
    
    dictionary.each do |dictionary_item|
      puts dictionary_item
      if word.downcase.include?(dictionary_item) # word in lowercase
        if words_hash.has_key?(dictionary_item)
          words_hash[dictionary_item] += 1
        else
          words_hash[dictionary_item] = 1
        end
      else
        next
      end

    end
  end

  # sorts the hash and returns a bidimensional rray with [[key, value]] format
  sorted_array_from_hash = words_hash.sort_by {|key, value| key}

  # creates a new sorted hash
  sorted_hash = Hash.new()

  sorted_array_from_hash.each do |element|
    sorted_hash[element[0]] = element[1]    
  end

  return sorted_hash

end

p substrings("Howdy partner, sit down! How's it going?",dictionary)

